### DISCUSSION TITLE
ENV vars are missing from actions context

### DISCUSSION THREAD

**Discussion by Troush** [community/community#26866] [Actions]
<p>TL:DR One of ENV vars is not dumping in github actions workflow, and I could not figure why.</p>
<p>Currently, I’m working on integrating with our scripts GH Actions, and one task is to hand over environment variables to our deployment scripts so that Ansible can use it<br>
​I made a short test, where I dump all the “env” context. For test purposes, I created the workflow, job, and step variables. Interestingly, when I dump the env context, the step variable is not included in the output.</p>
<p>GHA workflow example:</p>
<pre><code class="lang-auto"># Controls when the action will run. Triggers the workflow on push or pull request
# events but only for the master branch
on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]
    
# this is a workflow variable
env:
  WORKFLOW_VAR: "workflow variable"
  ANOTHER_WF_VAR: "another environment variable available to all jobs and steps in the workflow"
# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "build"
  build:
    # The type of runner that the job will run on
    runs-on: build
    
    # this is a job variable
    env:
      JOB_VAR: "job variable"
      ANOTHER_JOB_VAR: "another environment variable that is available only to all steps in build job"    
    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
    # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
    - uses: actions/checkout@v2
    # Runs a single command using the runners shell
    - name: Run a one-line script
      # this is a step variable
      env:
        STEP_VAR: "step variable"
        ANOTHER_STEP_VAR: "another environment variable that is available only to the step job"
      run: echo Hello, $STEP_VAR!
    # Runs a set of commands using the runners shell
    - name: Run a multi-line script
      run: |
        echo Add other actions to build,
        echo test, and deploy your project.
        echo $ANOTHER_JOB_VAR
        uname -a
    - name: Dump env context
      env:
        STEP_VAR_FOR_ENV_DUMP: "this is a step var for the env context dump step"
        ENV_CONTEXT: ${{ toJson(env) }}
      run: echo "$ENV_CONTEXT"
    - name: Dump GitHub context
      env:
        STEP_VAR_FOR_DUMP: "step var only for the GitHub context dump step"
        GITHUB_CONTEXT: ${{ toJson(github) }}
      run: echo "$GITHUB_CONTEXT"
</code></pre>
<p>Run results:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/c/ca25a4dc2cab2406e120891f0e226ec4500009c4.png" data-download-href="/uploads/short-url/sQh6fQhp2xiGOjBf6jfvSZaok96.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/1163074/181100367-808e9bdb-0019-480e-bed0-47b6daf41c5d.png" alt="image" data-base62-sha1="sQh6fQhp2xiGOjBf6jfvSZaok96" width="690" height="399" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/c/ca25a4dc2cab2406e120891f0e226ec4500009c4_2_690x399.png, https://aws1.discourse-cdn.com/github/original/2X/c/ca25a4dc2cab2406e120891f0e226ec4500009c4.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/c/ca25a4dc2cab2406e120891f0e226ec4500009c4.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/c/ca25a4dc2cab2406e120891f0e226ec4500009c4_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">909×526 35.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Dump env context step, should contain <code>STEP_VAR_FOR_ENV_DUMP</code>, but it does not.<br>
Any ideas why?</p>

**Comment by wei** [comment]
<p>I believe it’s because the new environment variable defined in the step is not yet in env context at time of execution <code>toJson(env)</code>.</p>
<p>You can try <code>echo "$STEP_VAR_FOR_ENV_DUMP"</code> instead and I bet that will work.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/troush">@Troush</a>,</p>
<p>Glad to see you in Github Community Forum!</p>
<p>The step level env only works in that step where it defines, it will not be added into GITHUB_CONTEXT.<br>
But you can find it in ‘env’ setting, and use <code>$varname</code> to invoke it.<br>
<img src="https://user-images.githubusercontent.com/52772761/181100368-9663032a-8a7c-41c7-a144-87e1dfd2f26e.png" alt="image" data-base62-sha1="8dJA79LHusWQ5KjguEH2BEGqSJ2" width="497" height="69"></p>
<p>To add env to Github Context in step, please use “<a href="https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable" rel="noopener nofollow ugc">set-env</a>” command:</p>
<pre><code class="lang-auto">    - name: add env var to github context
      run: |
          echo "::set-env name=action_state::yellow"
    - name: Dump env context for check
      env:
        ENV_CONTEXT: ${{ toJson(env) }}
      run: |
         echo "$ENV_CONTEXT"
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/8/82e8a6bace778e6fa93ebb83e8e61dddf21b5473.png" data-download-href="/uploads/short-url/iG4wvc0KqEW4Skb5ilOpyGnE31V.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/52772761/181100369-6a916cb1-1eab-4379-80da-d6b397648a0d.png" alt="image" data-base62-sha1="iG4wvc0KqEW4Skb5ilOpyGnE31V" width="690" height="213" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/8/82e8a6bace778e6fa93ebb83e8e61dddf21b5473_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">716×222 12 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Thanks</p>

**Reply by eguenou** [reply]
Hi,
This method will throw you the error below nowadays:

```
Error: Unable to process command '::set-env name=VAR_NAME::' successfully.
Error: The `set-env` command is disabled. Please upgrade to using Environment Files or opt into unsecure command execution by setting the `ACTIONS_ALLOW_UNSECURE_COMMANDS` environment variable to `true`. For more information see: https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/
```

They recommend this syntax:

`
echo "VAR_NAME=VAR_VALUE" >> "$GITHUB_ENV"
echo "action_state=yellow" >> "$GITHUB_ENV"
`

Source: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#environment-files
and https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/
